File: src/utils/mentions.ts
// Mention resolution utilities
// Converts @Name patterns to Fizzy ActionText mention attachments

import type { FizzyClient } from '../client';

export interface Mentionable {
  name: string;
  firstName: string;
  sgid: string;
  avatarUrl: string;
  userId: string;
}

/**
 * Fetch mentionable users from the /prompts/users endpoint.
 * Returns HTML with <lexxy-prompt-item> elements containing sgid and search attributes.
 */
export async function fetchMentionableUsers(client: FizzyClient): Promise<Mentionable[]> {
  const html = await client.getHtml('/prompts/users');
  const mentionables: Mentionable[] = [];

  // Parse <lexxy-prompt-item search="Name Initials" sgid="..."> elements
  const itemRegex = /<lexxy-prompt-item[^>]*>/gi;
  let match;
  while ((match = itemRegex.exec(html)) !== null) {
    const tag = match[0];

    const sgidMatch = tag.match(/sgid="([^"]+)"/);
    const searchMatch = tag.match(/search="([^"]+)"/);

    if (!sgidMatch || !searchMatch) continue;

    const sgid = sgidMatch[1];
    const search = searchMatch[1];

    // search is "Full Name Initials" — extract name (everything before last word which is initials)
    const parts = search.split(/\s+/);
    // Last part is initials (e.g., "WS"), rest is the name
    const name = parts.length > 1 ? parts.slice(0, -1).join(' ') : parts[0];
    const firstName = parts[0];

    // Extract avatar URL from inner img tag if present
    const avatarRegex = new RegExp(
      `<lexxy-prompt-item[^>]*sgid="${sgid.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}"[^>]*>[\\s\\S]*?<img[^>]*src="([^"]+)"`,
      'i'
    );
    const avatarMatch = html.match(avatarRegex);
    const avatarUrl = avatarMatch ? avatarMatch[1] : '';

    // Extract user ID from avatar URL (pattern: /ACCOUNT/users/USER_ID/avatar)
    const userIdMatch = avatarUrl.match(/\/users\/(\d+)\/avatar/);
    const userId = userIdMatch ? userIdMatch[1] : '';

    mentionables.push({ name, firstName, sgid, avatarUrl, userId });
  }

  return mentionables;
}

/**
 * Resolve @Name patterns in HTML to Fizzy mention attachments.
 *
 * Patterns:
 *   @firstName     — matches by first name (case-insensitive), skips if ambiguous
 *   @"Full Name"   — quoted, matches full name exactly
 *
 * Emails (user@example.com) are not treated as mentions.
 * Unresolved @Name stays as plain text with a warning to stderr.
 */
export function resolveMentions(html: string, mentionables: Mentionable[]): string {
  if (mentionables.length === 0) return html;

  // Collect all @mention matches with their positions, then replace from end to start
  interface MentionMatch {
    start: number;
    end: number;
    mentionable: Mentionable;
    original: string;
  }

  const matches: MentionMatch[] = [];

  // Pattern 1: @"Full Name" (quoted)
  const quotedRegex = /@"([^"]+)"/g;
  let m;
  while ((m = quotedRegex.exec(html)) !== null) {
    // Skip if inside an HTML tag
    if (isInsideTag(html, m.index)) continue;

    const fullName = m[1];
    const mentionable = mentionables.find(
      u => u.name.toLowerCase() === fullName.toLowerCase()
    );

    if (mentionable) {
      matches.push({
        start: m.index,
        end: m.index + m[0].length,
        mentionable,
        original: m[0],
      });
    } else {
      console.warn(`Warning: Could not resolve mention ${m[0]}`);
    }
  }

  // Pattern 2: @word (unquoted first name or full name)
  // Must NOT be preceded by a word character (to avoid matching emails like user@example.com)
  const unquotedRegex = /(?<![a-zA-Z0-9_.])@([a-zA-Z][a-zA-Z0-9_]*)/g;
  while ((m = unquotedRegex.exec(html)) !== null) {
    // Skip if inside an HTML tag
    if (isInsideTag(html, m.index)) continue;

    // Skip if this position already matched a quoted mention
    if (matches.some(existing => m!.index >= existing.start && m!.index < existing.end)) continue;

    const name = m[1];

    // Try exact first-name match (case-insensitive)
    const firstNameMatches = mentionables.filter(
      u => u.firstName.toLowerCase() === name.toLowerCase()
    );

    if (firstNameMatches.length === 1) {
      matches.push({
        start: m.index,
        end: m.index + m[0].length,
        mentionable: firstNameMatches[0],
        original: m[0],
      });
    } else if (firstNameMatches.length > 1) {
      console.warn(
        `Warning: Ambiguous mention ${m[0]} — matches: ${firstNameMatches.map(u => u.name).join(', ')}. Use @"Full Name" to disambiguate.`
      );
    } else {
      // Try full name match (single word matching first part)
      const fullNameMatch = mentionables.find(
        u => u.name.toLowerCase() === name.toLowerCase()
      );
      if (fullNameMatch) {
        matches.push({
          start: m.index,
          end: m.index + m[0].length,
          mentionable: fullNameMatch,
          original: m[0],
        });
      } else {
        console.warn(`Warning: Could not resolve mention ${m[0]}`);
      }
    }
  }

  // Sort matches by position descending so replacements don't shift indices
  matches.sort((a, b) => b.start - a.start);

  let result = html;
  for (const match of matches) {
    const mentionHtml = buildMentionHtml(match.mentionable);
    result = result.slice(0, match.start) + mentionHtml + result.slice(match.end);
  }

  return result;
}

/**
 * Check if a position in the string is inside an HTML tag (between < and >).
 */
function isInsideTag(html: string, pos: number): boolean {
  // Look backwards for < or >
  for (let i = pos - 1; i >= 0; i--) {
    if (html[i] === '>') return false;
    if (html[i] === '<') return true;
  }
  return false;
}

/**
 * Build the Fizzy mention HTML for a user.
 */
function buildMentionHtml(user: Mentionable): string {
  const avatarSrc = user.avatarUrl || `/${user.userId}/users/${user.userId}/avatar`;
  return (
    `<action-text-attachment sgid="${user.sgid}" content-type="application/vnd.actiontext.mention">` +
    `<img title="${user.name}" src="${avatarSrc}" width="48" height="48">` +
    `\n${user.firstName}` +
    `</action-text-attachment>`
  );
}

File: internal/commands/mentions.go
package commands

import (
	"fmt"
	"os"
	"regexp"
	"strings"
	"sync"
	"unicode"

	"github.com/basecamp/fizzy-cli/internal/client"
)

// mentionUser represents a mentionable user parsed from the /prompts/users endpoint.
type mentionUser struct {
	FirstName string // e.g. "Wayne"
	FullName  string // e.g. "Wayne Smith"
	SGID      string // signed global ID for ActionText
	AvatarSrc string // e.g. "/6103476/users/03f5awg7.../avatar"
}

// Package-level cache: populated once per CLI invocation.
var (
	mentionUsers []mentionUser
	mentionOnce  sync.Once
	mentionErr   error
)

// resetMentionCache resets the cache for testing.
func resetMentionCache() {
	mentionOnce = sync.Once{}
	mentionUsers = nil
	mentionErr = nil
}

// mentionRegex matches @Name patterns not preceded by word characters or dots
// (to avoid matching emails like user@example.com).
var mentionRegex = regexp.MustCompile(`(?:^|[^a-zA-Z0-9_.])@([a-zA-Z][a-zA-Z0-9_]*)`)

// promptItemRegex parses <lexxy-prompt-item> tags from the /prompts/users HTML.
var promptItemRegex = regexp.MustCompile(`<lexxy-prompt-item\s+search="([^"]+)"\s+sgid="([^"]+)"[^>]*>`)

// avatarRegex extracts the src attribute from the first <img> in editor template.
var avatarRegex = regexp.MustCompile(`<img[^>]+src="([^"]+)"`)

// resolveMentions scans text for @FirstName patterns and replaces them with
// ActionText mention HTML. If the text contains no @ characters, it is returned
// unchanged. On any error fetching users, the original text is returned with a
// warning printed to stderr.
func resolveMentions(text string, c client.API) string {
	if !strings.Contains(text, "@") {
		return text
	}

	mentionOnce.Do(func() {
		mentionUsers, mentionErr = fetchMentionUsers(c)
	})

	if mentionErr != nil {
		fmt.Fprintf(os.Stderr, "Warning: could not fetch mentionable users: %v\n", mentionErr)
		return text
	}

	if len(mentionUsers) == 0 {
		return text
	}

	// Find all @Name matches with positions
	type mentionMatch struct {
		start int // start of @Name (the @ character)
		end   int // end of @Name
		name  string
	}

	allMatches := mentionRegex.FindAllStringSubmatchIndex(text, -1)
	var matches []mentionMatch
	for _, loc := range allMatches {
		// loc[2]:loc[3] is the capture group (the name without @)
		nameStart := loc[2]
		nameEnd := loc[3]
		// The @ is one character before the name
		atStart := nameStart - 1
		name := text[nameStart:nameEnd]
		matches = append(matches, mentionMatch{start: atStart, end: nameEnd, name: name})
	}

	// Process from end to start so replacements don't shift indices
	for i := len(matches) - 1; i >= 0; i-- {
		m := matches[i]

		// Find matching user by first name (case-insensitive)
		var found []mentionUser
		for _, u := range mentionUsers {
			if strings.EqualFold(u.FirstName, m.name) {
				found = append(found, u)
			}
		}

		switch len(found) {
		case 1:
			html := buildMentionHTML(found[0])
			text = text[:m.start] + html + text[m.end:]
		case 0:
			fmt.Fprintf(os.Stderr, "Warning: could not resolve mention @%s\n", m.name)
		default:
			names := make([]string, len(found))
			for j, u := range found {
				names[j] = u.FullName
			}
			fmt.Fprintf(os.Stderr, "Warning: ambiguous mention @%s — matches: %s\n", m.name, strings.Join(names, ", "))
		}
	}

	return text
}

// fetchMentionUsers fetches the list of mentionable users from the API.
func fetchMentionUsers(c client.API) ([]mentionUser, error) {
	resp, err := c.GetHTML("/prompts/users")
	if err != nil {
		return nil, err
	}
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("unexpected status %d from /prompts/users", resp.StatusCode)
	}
	return parseMentionUsers(resp.Body), nil
}

// parseMentionUsers extracts mentionable users from the /prompts/users HTML.
// Each user is represented as a <lexxy-prompt-item> element with search and sgid
// attributes, containing <img> tags with avatar URLs.
func parseMentionUsers(html []byte) []mentionUser {
	htmlStr := string(html)
	items := promptItemRegex.FindAllStringSubmatch(htmlStr, -1)
	if len(items) == 0 {
		return nil
	}

	var users []mentionUser
	for _, item := range items {
		search := strings.TrimSpace(item[1])
		sgid := item[2]

		if search == "" || sgid == "" {
			continue
		}

		// Parse name from search attribute.
		// Format: "Full Name INITIALS [me]"
		// Strip trailing "me" and all-uppercase words (initials like "WS", "FMA").
		words := strings.Fields(search)
		for len(words) > 1 {
			last := words[len(words)-1]
			if last == "me" || isAllUpper(last) {
				words = words[:len(words)-1]
			} else {
				break
			}
		}

		fullName := strings.Join(words, " ")
		firstName := words[0]

		// Extract avatar URL: find the <img> after this sgid in the HTML.
		avatarSrc := ""
		sgidIdx := strings.Index(htmlStr, `sgid="`+sgid+`"`)
		if sgidIdx >= 0 {
			// Search for the first <img> after the sgid
			remainder := htmlStr[sgidIdx:]
			if m := avatarRegex.FindStringSubmatch(remainder); len(m) > 1 {
				avatarSrc = m[1]
			}
		}

		users = append(users, mentionUser{
			FirstName: firstName,
			FullName:  fullName,
			SGID:      sgid,
			AvatarSrc: avatarSrc,
		})
	}

	return users
}

// buildMentionHTML creates the ActionText attachment HTML for a mention.
func buildMentionHTML(u mentionUser) string {
	return fmt.Sprintf(
		`<action-text-attachment sgid="%s" content-type="application/vnd.actiontext.mention">`+
			`<img title="%s" src="%s" width="48" height="48">%s`+
			`</action-text-attachment>`,
		u.SGID, u.FullName, u.AvatarSrc, u.FirstName,
	)
}

// isAllUpper returns true if s is non-empty and all uppercase letters.
func isAllUpper(s string) bool {
	if s == "" {
		return false
	}
	for _, r := range s {
		if !unicode.IsUpper(r) {
			return false
		}
	}
	return true
}
